### PR TITLE
Generalize file splitting for output writers

### LINKED ISSUE #2967: Splitting NetCDF output with NetCDFOutputWriter
Opening this request as per the Slack conversation.  Sometimes file sizes become very big (and thus are not easy to transfer, etc.), and so quite often it would probably be good to split data files according to time intervals.

So something with API like
```
split_file = Filesize(10Mb)

split_file = TimeInterval(10days)
```

Is there any specific things that should be noted if we are to extend the `start_next_file()` functionality to NetCDF as well instead of just JLD2?  Like any flags or things that I should note.

### PR DESCRIPTION
**Pull request by glwagner**
This PR generalizes file splitting for `JLD2OutputWriter`, replacing `max_filesize` with a `file_splitting` keyword argument. To implement a maximum filesize we use

```julia
JLD2OutputWriter(model, outputs; file_splitting=FileSizeLimit(max_filesize), ...)
```

This feature should also be extended to `NetCDFOutputWriter` in this PR -- hopefully @josuemtzmo can help!

This PR will enable `TimeInterval` (or any other schedule --- `IterationInterval`, `WallTimeInterval`) to be used for splitting output files.

Closes #2967

### CODE REVIEW

File: src/OutputWriters/jld2_output_writer.jl [hunk 1]
```diff
@@ -80,10 +88,12 @@ Keyword arguments
...
 
-- `max_filesize`: The writer will stop writing to the output file once the file size exceeds `max_filesize`,
-                  and write to a new one with a consistent naming scheme ending in `part1`, `part2`, etc.
-                  Defaults to `Inf`.
-
+- `file_splitting`: Schedule for splitting the output file. The new files will be suffixed with
+                    `_part1`, `_part2`, etc. For example `file_splitting = FileSizeLimit(sz)` will
+                    split the output file when its size exceeds `sz`. Another example is 
+                    `file_splitting = TimeInterval(30days)`, which will split files every 30 days of
+                    simulation time. The default incurs no splitting.
```
Comment (navidcy) [inline src/OutputWriters/jld2_output_writer.jl:95]:
```suggestion
                    simulation time. The default incurs no splitting (`NoFileSplitting()`).
```

**Comment by navidcy** [comment]
We should also add a test similar to the one that was added in #3512.

**Comment by glwagner** [comment]
> We should also add a test similar to the one that was added in #3512.

Doesn't that already exist? 

We just have to adapt the test

**Comment by glwagner** [comment]
Also open to comments about the new names / kwargs.

**Comment by navidcy** [comment]
> > We should also add a test similar to the one that was added in #3512.
> 
> Doesn't that already exist?
> 
> We just have to adapt the test

Yes, the test I think only tests the `NetCDFOutputWriter`.

**Comment by josuemtzmo** [comment]
> > Doesn't that already exist?
> > We just have to adapt the test
> 
> Yes, the test I think only tests the `NetCDFOutputWriter`.

There is already a test for the `JLD2OutputWriter`

**Comment by josuemtzmo** [comment]
I will work on the time splitting. Should we do it in this PR? or should I do another one once it is ready?

File: src/OutputWriters/jld2_output_writer.jl [hunk 2]
```diff
@@ -10,7 +10,10 @@ default_included_properties(::ShallowWaterModel) = [:grid, :coriolis, :closure]
 default_included_properties(::HydrostaticFreeSurfaceModel) = [:grid, :coriolis, :buoyancy, :closure]
 
 update_file_splitting_schedule!(schedule, filepath) = nothing
-update_file_splitting_schedule!(schedule::FileSizeLimit, filepath) = schedule.path = filepath
+
+function update_file_splitting_schedule!(schedule::FileSizeLimit, filepath) 
+    schedule.path = filepath
```
Comment (glwagner) [inline src/OutputWriters/jld2_output_writer.jl]:
```suggestion
    schedule.path = filepath
    return nothing
```

We use explicit returns (if the function is not just one line

File: src/OutputWriters/jld2_output_writer.jl [hunk 3]
```diff
@@ -258,18 +261,17 @@ end
...
+    if iteration_exists(writer.filepath, current_iteration)
 
         if writer.overwrite_existing
             # Something went wrong, so we remove the file and re-initialize it.
-            rm(path, force=true)
+            rm(writer.filepath, force=true)
             initialize_jld2_file!(writer, model)
         else # nothing we can do since we were asked not to overwrite_existing, so we skip output writing
-            @warn "Iteration $current_iteration was found in $path. Skipping output writing (for now...)"
+            @warn "Iteration $current_iteration was found in $writer.filepath. Skipping output writing (for now...)"
```
Comment (glwagner) [inline src/OutputWriters/jld2_output_writer.jl:263]:
```suggestion
            @warn "Iteration $current_iteration was found in $(writer.filepath). Skipping output writing (for now...)"
```

File: src/OutputWriters/jld2_output_writer.jl [hunk 4]
```diff
@@ -258,18 +261,17 @@ end
 function write_output!(writer::JLD2OutputWriter, model)
 
     verbose = writer.verbose
-    path = writer.filepath
```
Comment (glwagner) [inline src/OutputWriters/jld2_output_writer.jl:250]: why this change?

File: test/test_jld2_output_writer.jl [hunk 5]
```diff
@@ -58,7 +60,7 @@ function test_jld2_file_splitting(arch)
                           including = [:grid],
                           array_type = Array{Float64},
                           with_halos = true,
-                          file_splitting = FileSizeLimit(200KiB),
+                          file_splitting = filesizelimit,
```
Comment (glwagner) [inline test/test_jld2_output_writer.jl:63]:
```suggestion
                          file_splitting = file_size_limit,
```

I like how it was previously though, faster to read...

File: src/OutputWriters/jld2_output_writer.jl (hunk 4 above)
Comment (josuemtzmo) [reply src/OutputWriters/jld2_output_writer.jl:250]: Because `path` was used later on in the code, but it wasn't updated. Thus the code crashed while  creating the file and when the `writer.filepath` changed  but not the `path` . The easier fix was to replace `path` by the `writer.filepath`.

File: src/OutputWriters/output_writer_utils.jl [hunk 6]
```diff
@@ -6,11 +6,52 @@ using Oceananigans.Fields: AbstractField, indices, boundary_conditions, instanti
...
+
+Return a schedule that actuates when the file at `path` exceeds
+the `size_limit`.
+
+The `path` is automatically added and updated when `FileSizeLimit` is
+used with an output writer, and should not be provided manually.
+"""
+FileSizeLimit(size_limit) = FileSizeLimit(size_limit, "")
+
+(fsl::FileSizeLimit)(model) = filesize(fsl.path) >= fsl.size_limit
```
Comment (navidcy) [inline src/OutputWriters/output_writer_utils.jl:31]: not necessary, I just find this easier/soother for the eye

```suggestion
(fsl::FileSizeLimit)(model) = filesize(fsl.path) ≥ fsl.size_limit
```

File: test/test_jld2_output_writer.jl [hunk 7]
```diff
@@ -49,7 +49,7 @@ function test_jld2_file_splitting(arch)
     function fake_bc_init(file, model)
         file["boundary_conditions/fake"] = π
     end
-
+
```
Comment (navidcy) [inline test/test_jld2_output_writer.jl:52]:
```suggestion

```

**Review by navidcy — APPROVED** [review]
lgtm!

**Comment by josuemtzmo** [comment]
Why are some one the test failing?

**Comment by navidcy** [comment]
@glwagner would you like to have a look at this or should we merge when tests pass?

**Comment by navidcy** [comment]
Some of the doctests are failing because of the changes:

https://buildkite.com/clima/oceananigans/builds/15164#018e6aa7-67f8-4280-b331-ad7179892dd1/32-18831

File: src/OutputWriters/jld2_output_writer.jl (hunk 4 above)
Comment (glwagner) [reply src/OutputWriters/jld2_output_writer.jl:250]: Ok, that's a good reason!

**Comment by glwagner** [comment]
I think it looks pretty good! Thanks @josuemtzmo for the work on `NetCDFOutputWriter`.

In a future PR, we can add a test to ensure that `file_splitting = TimeInterval(T)` works.

Also I'm trying to brainstorm a better word to use than "actuated" for talking about scheduling.

But I think once the tests pass we should merge this.

File: src/OutputWriters/netcdf_output_writer.jl [hunk 8]
```diff
@@ -577,11 +585,16 @@ dictify(outputs::LagrangianParticles) = Dict("particles" => outputs)
...
+#####
+
 function start_next_file(model, ow::NetCDFOutputWriter)
     verbose = ow.verbose
-    sz = filesize(ow.filepath)
+
     verbose && @info begin
-        "Filesize $(pretty_filesize(sz)) has exceeded maximum file size $(pretty_filesize(ow.max_filesize))."
+        schedule_type = summary(ow.file_splitting)
+        "Splitting output because $(schedule_type) has been actuated."
```
Comment (glwagner) [inline src/OutputWriters/netcdf_output_writer.jl:597]:
```suggestion
        "Splitting output because $(schedule_type) is activated."
```

File: src/OutputWriters/jld2_output_writer.jl [hunk 9]
```diff
@@ -311,9 +312,10 @@ end
 
 function start_next_file(model, writer::JLD2OutputWriter)
     verbose = writer.verbose
-    sz = filesize(writer.filepath)
+   
     verbose && @info begin
-        "Filesize $(pretty_filesize(sz)) has exceeded maximum file size $(pretty_filesize(writer.max_filesize))."
+        schedule_type = summary(writer.file_splitting)
+        "Splitting output because $(schedule_type) has been actuated."
```
Comment (glwagner) [inline src/OutputWriters/jld2_output_writer.jl:318]:
```suggestion
        "Splitting output because $(schedule_type) is activated."
```
